### DISCUSSION TITLE
CreateRule using other collection

### DISCUSSION THREAD

**Discussion by lbensaad** [pocketbase/pocketbase#5551] [Q&A] [answered]
Hi,
I have collections  `card(id, owner, balance)` and `tx(id, from_card, to_card, amount)`
How to set the create rule for `tx` so that only the `owner` of `from_card` can create a `tx`, and the submitted `amount` data should be less than the `from_card.balance`

**Comment by ganigeorgiev** [comment]
If we assume that `from_card` is a single `relation` to `card` and that `owner` is a single `relation` pointing to the `users` (or other auth collection), then the following Create API rule should suffice:

```js
@request.auth.id != "" &&
@request.data.from_card.owner = @request.auth.id &&
@request.data.amount < @request.data.from_card.balance
```

If that doesn't help, please export your collections configuration as json (_Admin UI > Settings > Export collections_).

**Comment by lbensaad** [comment]
Thanks a lot @ganigeorgiev . I found the issue
To insure data integrity, I opted to use sql triggers to update cards balance after tx is inserted
I have created this sql trigger
```
CREATE TRIGGER update_cards_balance
AFTER INSERT ON txs FOR EACH ROW
BEGIN
	update cards
		set balance = balance - new.amount
			WHERE id = new.from_card;
	update cards
		set balance = balance + new.amount
			WHERE id = new.to_card;
END;
```
I started card 1 balance = 100, when I tested submitting this tx multiple times ```{ from_card: '1', to_card: '2', amount: 15}```. 
It works fine till the balance is 25 i get error 400, and the logs table i got this error:
```
data.error: Failed to create record.
data.details: DrySubmit create rule failure: sql: no rows in result set
```
I don't know how the createRule is working but i think when balance is 25, the sql trigger reduces it to 10, then the create rule 
filter ```@request.data.amount < @request.data.from_card.balance``` will not find any tx hence the error `no rows in result set` and the sql  transaction will not be committed and it will be rolled back, because after the error i found the balance is still 25.
is this explanation correct? or what do you the the issue is?

**Reply by ganigeorgiev** [reply, marked as answer]
The Create API rule executes **after** a dry submit of the record, aka. when the rule executes the record is already inserted in the database and it will fire your trigger.

I don't recommend using SQLite triggers as they will not fire the app event hooks and the realtime subscription for the created and updated records will not work. As an alternative you could use the app event hooks. Here is one possible example using the JSVM `pb_hooks` (_I haven't tested it_):

```js
// pb_hooks/main.pb.js

/// <reference path="../pb_data/types.d.ts" />
onModelAfterCreate((e) => {
    const tx = e.model

    try {
        $app.dao().runInTransaction((txDao) => {
            const fromCard = $app.dao().findRecordById("card", tx.get("from_card"))
            fromCard.set("balance", fromCard.get("balance") - tx.get("amount"))
            txDao.saveRecord(fromCard)

            const toCard = $app.dao().findRecordById("card", tx.get("to_card"))
            toCard.set("balance", toCard.get("balance") + tx.get("amount"))
            txDao.saveRecord(toCard)
        })
    } catch(err) {
        // in case the transaction fails try to manually delete the just created record
        // or maybe updated it as "unprocessed/failed"? 
        $app.dao().deleteRecord(tx)
        throw err // rethrow
    }
}, "tx")
```

Note that currently there is no mechanism to wrap the entire request operation + the custom hooks in a single transaction. This will be possible with the next v0.23.0 release soon (_there are no ETAs yet_). The above could look something like:

```js
// pb_hooks/main.v0230.pb.js

/// <reference path="../pb_data/types.d.ts" />
onRecordCreate((e) => {
    const originalApp = e.app
    e.app.runInTransaction((txApp) => {
        // temp replace the event app with the transactional one
        e.app = txApp;

        // execute the original create action
        e.next()

        const fromCard = e.app.findRecordById("card", e.record.get("from_card"))
        fromCard.set("balance", fromCard.get("balance") - e.record.get("amount"))
        e.app.save(fromCard) // this also runs the model validations

        const toCard = e.app.findRecordById("card", e.record.get("to_card"))
        toCard.set("balance", toCard.get("balance") + e.record.get("amount"))
        e.app.save(toCard) // this also runs the model validations
    })
    e.app = originalApp
}, "tx")
```

**Comment by lbensaad** [comment]
ok great, thanks a lot @ganigeorgiev

**Comment by lbensaad** [comment]
But why not execute the create api rule before the insert, shouldn't the rule work as a guard?

**Reply by ganigeorgiev** [reply]
Because rules are translated to plain SQL statement, aka. it requires the record to exist in the database in order the check to work.

The Create API rule behavior may change in the future as several other users also complained about it but I have to think a little more on it and for now is pushed aside.

**Reply by ganigeorgiev** [reply]
_Just in case someone stumble here: the dry submit behavior is discussed to change in v0.24+ per https://github.com/pocketbase/pocketbase/discussions/6073._

**Comment by lbensaad** [comment]
In my opinion, the create rule should act as an access control, after it is run it should use the List/View Rules to filter the returned value;
For now i kept the sql trigger and removed the CreateRule
